src/vmm/src/gdb/target.rs

```diff
@@ -0,0 +1,682 @@
+        }
+
+        self.paused_vcpu = None;
+
+        Ok(())
+    }
+
+    /// Resets all Vcpus to their base state
+    fn reset_all_states(&mut self) {
+        for value in self.vcpu_state.iter_mut() {
+            value.reset_vcpu_state();
+        }
+    }
+
+    /// Shuts down the VMM
+    pub fn shutdown(&self) {
+        self.vmm
+            .lock()
+            .expect("error unlocking vmm")
+            .stop(FcExitCode::Ok)
+    }
+
+    /// Pauses the requested Vcpu
+    pub fn request_pause(&mut self, tid: Tid) -> Result<(), Error> {
+        let vcpu_state = &mut self.vcpu_state[tid_to_vcpuid(tid)];
+
+        if vcpu_state.paused {
+            info!("Attempted to pause a vcpu already paused.");
+            // Pausing an already paused vcpu is not considered an error case from GDB
+            return Ok(());
+        }
+
+        let cpu_handle = &self.vmm.lock()?.vcpus_handles[tid_to_vcpuid(tid)];
+
+        cpu_handle.send_event(VcpuEvent::Pause)?;
+        let _ = cpu_handle.response_receiver().recv()?;
+
+        vcpu_state.paused = true;
+        Ok(())
+    }
+
+    /// A helper function to allow the event loop to inject this breakpoint back into the Vcpu
+    pub fn inject_bp_to_guest(&mut self, tid: Tid) -> Result<(), Error> {
+        let vcpu_state = &mut self.vcpu_state[tid_to_vcpuid(tid)];
+        kvm_debug::vcpu_inject_bp(&vcpu_state.vcpu_fd, &self.hw_breakpoints, false)
+    }
+
+    /// Resumes the Vcpu, will return early if the Vcpu is already running
+    pub fn request_resume(&mut self, tid: Tid) -> Result<(), Error> {
+        let vcpu_state = &mut self.vcpu_state[tid_to_vcpuid(tid)];
+
+        if !vcpu_state.paused {
+            info!("Attempted to resume a vcpu already running.");
+            // Resuming an already running Vcpu is not considered an error case from GDB
+            return Ok(());
+        }
+
+        let cpu_handle = &self.vmm.lock()?.vcpus_handles[tid_to_vcpuid(tid)];
+        cpu_handle.send_event(VcpuEvent::Resume)?;
+
+        let response = cpu_handle.response_receiver().recv()?;
+        if let VcpuResponse::NotAllowed(message) = response {
+            error!("Response resume : {message}");
+            return Err(Error::VcuRequestError);
+        }
+
+        vcpu_state.paused = false;
+        Ok(())
+    }
+
+    /// Identifies why the specifc core was paused to be returned to GDB if None is returned this
+    /// indicates to handle this internally and don't notify GDB
+    pub fn get_stop_reason(&self, tid: Tid) -> Result<Option<BaseStopReason<Tid, u64>>, Error> {
+        let vcpu_state = &self.vcpu_state[tid_to_vcpuid(tid)];
+        if vcpu_state.single_step {
+            return Ok(Some(MultiThreadStopReason::SignalWithThread {
+                tid,
+                signal: Signal::SIGTRAP,
+            }));
+        }
+
+        let Ok(regs) = kvm_debug::get_regs(&vcpu_state.vcpu_fd) else {
+            return Ok(Some(MultiThreadStopReason::SwBreak(tid)));
+        };
```

Comment:
Is it expected for `get_regs` to error out in this case?

Reply:
It _could_ in theory error out,  so I've just handled that case

Reply:
But should we return `MultiThreadStopReason::SwBreak` in this case or smth else?

Reply:
Ok